### ISSUE TITLE
Pre-training Phase Implementation for 3D-LLaVA

### ISSUE THREAD

**Issue by ijliu** [djiajunustc/3D-LLaVA#1]
Thank you sincerely for open-sourcing the 3D-LLaVA project, which I find highly intriguing. I have successfully reproduced the codebase locally and completed the fine-tuning stage. However, I noticed that the repository does not include pre-training code. Could you kindly clarify the following key considerations for the pre-training phase ?

**Comment by djiajunustc** [comment]
Yes. You can reimplement the pretraining phase by modifying Oneformer3D: https://github.com/filaPro/oneformer3d.

**Comment by Djerry-h** [comment]
Hi，Was the Stage 1 model trained exclusively using the OneFormer3D framework to obtain the corresponding weight parameters?

**Comment by yt2639** [comment]
Hii @djiajunustc thanks for providing the code. I also have the same question. 

Since the paper mentions the pre-training stage uses `ScanNet200` for instance segmentation. Do we follow https://github.com/filaPro/oneformer3d?tab=readme-ov-file on:
```
ScanNet200
For ScanNet200 we present the model with [MinkowskiEngine](https://github.com/NVIDIA/MinkowskiEngine) backbone, superpoint pooling, selecting all queries, and predicting semantics directly from instance queries. Backbone is initialized from [Mask3D](https://github.com/JonasSchult/Mask3D) checkpoint. It should be [downloaded](https://github.com/oneformer3d/oneformer3d/releases/download/v1.0/mask3d_scannet200.pth) and put to work_dirs/tmp before training.

# train (with validation)
python tools/train.py configs/oneformer3d_1xb4_scannet200.py
# test
python tools/test.py configs/oneformer3d_1xb4_scannet200.py \
    work_dirs/oneformer3d_1xb4_scannet/epoch_512.pth
```
and in the meantime, use OpenScene's multi-view fused 2D features to jointly perform **Stage-1 Pre-training 3D Scene Encoder and OST**?

Thanks!

**Comment by djiajunustc** [comment]
Yes. You can first extract the vision encoder feature from the LLaVA model (it is actually CLIP-ViT-L). Then, fusing multi-view features at each point as OpenScene. Finally, performing joint training.

**Comment by yt2639** [comment]
Hi @djiajunustc , thanks for your response. I noticed OST is a new module proposed in your paper. And this pre-training stage involves both the 3D Encoder and the OST (if I understand it correctly). So we need to perform this pre-training using this code base (as in compose some code to train 3D Encoder+OST) instead of directly using OneFormer3D. Is this understanding correct?

Thanks!

**Comment by yt2639** [comment]
Hi @djiajunustc , I have managed to compose code for the pretraining stage using this and OneFormer3D's code bases. However, when I evaluate the pretrained model [provided in this repo](https://github.com/djiajunustc/3D-LLaVA/blob/main/checkpoints/pc_pretrained/ost-sa-only-llava-align-scannet200.pth) on **ScanNet200 Dataset**, the results seem worse than OneFormer3D's results. Specifically,
```
06/19 04:39:06 - mmengine - INFO -                                                                                                                                                        
+-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
| classes                     | AP_0.25 | AP_0.50 | AP     | Prec_0.50 | Rec_0.50 |                                                                                                       
+-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
| chair                       | 0.7714  | 0.6212  | 0.3851 | 0.7298    | 0.6884   |                                                                                                       
| table                       | 0.7431  | 0.5359  | 0.3311 | 0.5944    | 0.5441   |                                                                                                       
| door                        | 0.8033  | 0.6746  | 0.4470 | 0.8468    | 0.6144   |                                                                                                       
| couch                       | 0.8043  | 0.5794  | 0.3040 | 0.7681    | 0.5464   |                                                                                                       
| cabinet                     | 0.4361  | 0.3197  | 0.1736 | 0.4152    | 0.4465   |                                                                                                       
| shelf                       | 0.5548  | 0.3826  | 0.2185 | 0.4000    | 0.5325   |                                                                                                       
| desk                        | 0.7451  | 0.4393  | 0.1724 | 0.5385    | 0.5512   |                                                                                                       
| office chair                | 0.2024  | 0.1519  | 0.0956 | 0.2614    | 0.3810   |    
...
| power strip                 | nan     | nan     | nan    | nan       | nan      |          
| calendar                    | 0.9028  | 0.9028  | 0.8153 | 0.7500    | 1.0000   |          
| poster                      | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
| potted plant                | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
| luggage                     | nan     | nan     | nan    | nan       | nan      |          
| mattress                    | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
+-----------------------------+---------+---------+--------+-----------+----------+          
| Overall                     | 0.3397  | 0.2759  | 0.1832 | 0.4262    | 0.3800   |          
+-----------------------------+---------+---------+--------+-----------+----------+    
06/19 04:39:09 - mmengine - INFO - Epoch(test) [312/312]    
**miou: 0.1958**  all_ap: 0.1832  all_ap_50%: 0.2759  all_ap_25%: 0.3397  **pq: 0.1911**  data_time: 0.4529  time: 0.5743
```
OneFormer3D's results, obtained from their paper:
```
+-----------------------------+---------+---------+--------+-----------+----------+    
| OneFormer3D                 | 0.454   | 0.408   | 0.306  | ...     | ...    |          
+-----------------------------+---------+---------+--------+-----------+----------+    
miou: 0.301
PQ: 0.312
```
Is this result expected? Or this inferior result indicates a wrong implementation?

Thank you so much for your assistance!

**Comment by jliu4ai** [comment]
Hi @djiajunustc,

Thanks for the fantastic work! 

Would you mind sharing the pretraining code? It will definitely make a lot of sense for the community.

Thanks!

**Comment by Djerry-h** [comment]
> 嗨，我已经设法使用这个和 OneFormer3D 的代码库为预训练阶段编写代码。但是，当我在 **ScanNet200 数据集**上评估[此存储库中提供的](https://github.com/djiajunustc/3D-LLaVA/blob/main/checkpoints/pc_pretrained/ost-sa-only-llava-align-scannet200.pth)预训练模型时，结果似乎比 OneFormer3D 的结果更差。具体说来
> 
> ```
> 06/19 04:39:06 - mmengine - INFO -                                                                                                                                                        
> +-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
> | classes                     | AP_0.25 | AP_0.50 | AP     | Prec_0.50 | Rec_0.50 |                                                                                                       
> +-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
> | chair                       | 0.7714  | 0.6212  | 0.3851 | 0.7298    | 0.6884   |                                                                                                       
> | table                       | 0.7431  | 0.5359  | 0.3311 | 0.5944    | 0.5441   |                                                                                                       
> | door                        | 0.8033  | 0.6746  | 0.4470 | 0.8468    | 0.6144   |                                                                                                       
> | couch                       | 0.8043  | 0.5794  | 0.3040 | 0.7681    | 0.5464   |                                                                                                       
> | cabinet                     | 0.4361  | 0.3197  | 0.1736 | 0.4152    | 0.4465   |                                                                                                       
> | shelf                       | 0.5548  | 0.3826  | 0.2185 | 0.4000    | 0.5325   |                                                                                                       
> | desk                        | 0.7451  | 0.4393  | 0.1724 | 0.5385    | 0.5512   |                                                                                                       
> | office chair                | 0.2024  | 0.1519  | 0.0956 | 0.2614    | 0.3810   |    
> ...
> | power strip                 | nan     | nan     | nan    | nan       | nan      |          
> | calendar                    | 0.9028  | 0.9028  | 0.8153 | 0.7500    | 1.0000   |          
> | poster                      | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> | potted plant                | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> | luggage                     | nan     | nan     | nan    | nan       | nan      |          
> | mattress                    | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> +-----------------------------+---------+---------+--------+-----------+----------+          
> | Overall                     | 0.3397  | 0.2759  | 0.1832 | 0.4262    | 0.3800   |          
> +-----------------------------+---------+---------+--------+-----------+----------+    
> 06/19 04:39:09 - mmengine - INFO - Epoch(test) [312/312]    
> **miou: 0.1958**  all_ap: 0.1832  all_ap_50%: 0.2759  all_ap_25%: 0.3397  **pq: 0.1911**  data_time: 0.4529  time: 0.5743
> ```
> 
> OneFormer3D 从他们的论文中获得的结果：
> 
> ```
> +-----------------------------+---------+---------+--------+-----------+----------+    
> | OneFormer3D                 | 0.454   | 0.408   | 0.306  | ...     | ...    |          
> +-----------------------------+---------+---------+--------+-----------+----------+    
> miou: 0.301
> PQ: 0.312
> ```
> 
> 此结果是否符合预期？或者这个劣质结果表明实现错误？
> 
> 非常感谢您的帮助！

Would you mind sharing the  code? I want to work with you to finish this work.

**Comment by yukili2000** [comment]
> > 嗨，我已经设法使用这个和 OneFormer3D 的代码库为预训练阶段编写代码。但是，当我在 **ScanNet200 数据集**上评估[此存储库中提供的](https://github.com/djiajunustc/3D-LLaVA/blob/main/checkpoints/pc_pretrained/ost-sa-only-llava-align-scannet200.pth)预训练模型时，结果似乎比 OneFormer3D 的结果更差。具体说来
> > ```
> > 06/19 04:39:06 - mmengine - INFO -                                                                                                                                                        
> > +-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
> > | classes                     | AP_0.25 | AP_0.50 | AP     | Prec_0.50 | Rec_0.50 |                                                                                                       
> > +-----------------------------+---------+---------+--------+-----------+----------+                                                                                                       
> > | chair                       | 0.7714  | 0.6212  | 0.3851 | 0.7298    | 0.6884   |                                                                                                       
> > | table                       | 0.7431  | 0.5359  | 0.3311 | 0.5944    | 0.5441   |                                                                                                       
> > | door                        | 0.8033  | 0.6746  | 0.4470 | 0.8468    | 0.6144   |                                                                                                       
> > | couch                       | 0.8043  | 0.5794  | 0.3040 | 0.7681    | 0.5464   |                                                                                                       
> > | cabinet                     | 0.4361  | 0.3197  | 0.1736 | 0.4152    | 0.4465   |                                                                                                       
> > | shelf                       | 0.5548  | 0.3826  | 0.2185 | 0.4000    | 0.5325   |                                                                                                       
> > | desk                        | 0.7451  | 0.4393  | 0.1724 | 0.5385    | 0.5512   |                                                                                                       
> > | office chair                | 0.2024  | 0.1519  | 0.0956 | 0.2614    | 0.3810   |    
> > ...
> > | power strip                 | nan     | nan     | nan    | nan       | nan      |          
> > | calendar                    | 0.9028  | 0.9028  | 0.8153 | 0.7500    | 1.0000   |          
> > | poster                      | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> > | potted plant                | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> > | luggage                     | nan     | nan     | nan    | nan       | nan      |          
> > | mattress                    | 0.0000  | 0.0000  | 0.0000 | 0.0000    | 0.0000   |          
> > +-----------------------------+---------+---------+--------+-----------+----------+          
> > | Overall                     | 0.3397  | 0.2759  | 0.1832 | 0.4262    | 0.3800   |          
> > +-----------------------------+---------+---------+--------+-----------+----------+    
> > 06/19 04:39:09 - mmengine - INFO - Epoch(test) [312/312]    
> > **miou: 0.1958**  all_ap: 0.1832  all_ap_50%: 0.2759  all_ap_25%: 0.3397  **pq: 0.1911**  data_time: 0.4529  time: 0.5743
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > OneFormer3D 从他们的论文中获得的结果：
> > ```
> > +-----------------------------+---------+---------+--------+-----------+----------+    
> > | OneFormer3D                 | 0.454   | 0.408   | 0.306  | ...     | ...    |          
> > +-----------------------------+---------+---------+--------+-----------+----------+    
> > miou: 0.301
> > PQ: 0.312
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 此结果是否符合预期？或者这个劣质结果表明实现错误？
> > 非常感谢您的帮助！
> 
> Would you mind sharing the code? I want to work with you to finish this work.

你好，可以交流一下具体是怎么实现的吗

**Comment by sujiez** [comment]
well, I actually managed to assemble the eval pipeline (borrowed from Oneformer3D) and tested on the OST part with the pre-trained weight provided, and found nearly no substantial changes from the performance of Oneformer3D.

| classes | AP_0.25 | AP_0.50 | AP     | Prec_0.50 | Rec_0.50 |
Overall | 0.4585  | 0.4005  | 0.3021 | 0.5460    | 0.5053   |
+---------+---------+---------+--------+-----------+----------+

==== Instance segmentation (n=312) ====
  mAP    = 30.21
  AP@50  = 40.05
  AP@25  = 45.85

==== Semantic segmentation ====
  mIoU   = 29.16  (over 189 classes present in GT)

seems weird to me since this is only a cut down version of Oneformer3D with disturbing loss...
